File: web/api/schemas/market.py
"""Market and sector response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class MarketSnapshot(BaseModel):
    regime: str = Field(..., description="市场状态: bull/bear/sideways")
    raw_value: str = Field(default="unknown", description="未做稳定确认的原始市场状态")
    score: float = Field(default=50.0, description="Regime 评分 0-100")
    ma_trend: str = Field(..., description="均线趋势描述")
    volume_trend: str = Field(..., description="成交量趋势")
    breadth: float = Field(..., description="全市场上涨家数占比")
    breadth_detail: dict[str, Any] = Field(default_factory=dict, description="全市场宽度明细")
    score_components: dict[str, Any] = Field(default_factory=dict, description="Regime 评分分项")
    stability: dict[str, Any] = Field(default_factory=dict, description="Regime 稳定确认状态")
    confidence: float = Field(default=0.0, description="置信度 0-1")


class KLineItem(BaseModel):
    date: str
    open: float
    high: float
    low: float
    close: float
    volume: int


class MarketResponse(BaseModel):
    regime: MarketSnapshot
    kline: list[KLineItem]
    config: dict[str, Any]
    updated: str


class MultiAssetCard(BaseModel):
    key: str
    label: str
    symbol: str
    value: float | None = None
    change: float = 0.0
    change_pct: float = 0.0
    unit: str = ""
    series: list[dict[str, Any]] = []
    data_source: str = Field(default="real", description="real | proxy | placeholder | cached | missing")
    source_detail: str = Field(default="", description="Human-readable detail about the data origin")


class RegimeSnapshot(BaseModel):
    regime: str = "unknown"
    regime_score: float = 50.0
    multi_asset: list[MultiAssetCard] = []
    freshness: dict[str, Any] = {}
    updated: str = ""


class MarketRegimeResponse(BaseModel):
    regime: dict[str, Any] = {}
    multi_asset: list[dict[str, Any]] = []
    freshness: dict[str, Any] = {}
    config: dict[str, Any] = {}
    position_capacity: dict[str, Any] = {}
    updated: str = ""


class MarketOverviewResponse(BaseModel):
    regime: dict[str, Any] = {}
    kline: list[dict[str, Any]] = []
    range: str = "6M"
    multi_asset: list[dict[str, Any]] = []
    macro: list[dict[str, Any]] = []
    freshness: dict[str, Any] = {}
    pool_size: int = 0
    position_capacity: dict[str, Any] = {}
    config: dict[str, Any] = {}
    updated: str = ""


class SectorOverviewResponse(BaseModel):
    sectors: list[dict[str, Any]] = []
    total_sectors: int = 0
    top_performers: list[dict[str, Any]] = []
    bottom_performers: list[dict[str, Any]] = []
    signal_dispersion: float = 0.0
    data_source: str = ""
    capital_source: str = ""
    freshness: dict[str, Any] = {}

File: web/frontend/src/api/types/market.ts
export interface RegimeBreadthDetail {
  advance_ratio?: number;
  above_ma20?: number;
  above_ma60?: number;
  above_ma120?: number;
  sample_size?: number;
  up_count?: number;
  down_count?: number;
  unchanged_count?: number;
  as_of?: string;
}

export interface RegimeScoreComponents {
  trend?: number;
  breadth?: number;
  risk?: number;
  volume?: number;
  trend_raw?: number;
  breadth_raw?: number;
  risk_raw?: number;
  volume_raw?: number;
  sample_size?: number;
  [key: string]: number | undefined;
}

export interface RegimeStability {
  raw_value?: "bull" | "bear" | "sideways" | "unknown" | "";
  confirmed_value?: "bull" | "bear" | "sideways" | "unknown" | "";
  pending_value?: "bull" | "bear" | "sideways" | "unknown" | "";
  pending_count?: number;
  min_dwell?: number;
  confirmed_changed?: boolean;
  score?: number;
  as_of?: string;
}

export type RegimeDetectionMethod = "rule_based" | "hmm" | "hybrid";
export type RegimeProbabilityMap = Partial<Record<"bull" | "sideways" | "bear", number>>;

export interface RegimeData {
  value: "bull" | "bear" | "sideways" | "unknown";
  raw_value?: "bull" | "bear" | "sideways" | "unknown";
  score?: number;
  ma_trend: string;
  volume_trend: string;
  breadth: number;
  breadth_detail?: RegimeBreadthDetail;
  score_components?: RegimeScoreComponents;
  stability?: RegimeStability;
  regime_probs?: RegimeProbabilityMap;
  detection_method?: RegimeDetectionMethod;
  hmm_confidence?: number;
  hmm_entropy?: number;
  decision_reason?: string;
}

export interface KlinePoint {
  date: string;
  open: number;
  close: number;
  high: number;
  low: number;
  volume: number;
}

export interface MarketSeriesPoint {
  date: string;
  value: number;
}

export interface MarketAssetCard {
  key: string;
  label: string;
  symbol: string;
  value: number | null;
  change: number;
  change_pct: number;
  unit: string;
  series: MarketSeriesPoint[];
  data_source: string;       // "real" | "proxy" | "placeholder" | "cached" | "missing"
  source_detail?: string;    // human-readable detail
}

export interface MacroCard {
  key: string;
  label: string;
  value: number | null;
  prev: number | null;
  unit: string;
  date: string;
  series: MarketSeriesPoint[];
}

export interface MarketAlert {
  level: "success" | "warning" | "danger" | "info";
  title: string;
  detail: string;
  time: string;
}

export interface PositionCapacity {
  current: number;
  max: number;
}

export interface MarketResponse {
  regime: RegimeData;
  kline: KlinePoint[];
  range?: string;
  multi_asset?: MarketAssetCard[];
  macro?: MacroCard[];
  freshness?: { market: string; macro: string };
  pool_size?: number;
  position_capacity?: PositionCapacity;
  config?: Record<string, any>;
  updated?: string;
}

export interface RegimeResponse {
  regime: RegimeData;
  multi_asset: MarketAssetCard[];
  freshness: { market: string };
  updated: string;
  config?: Record<string, any>;
  position_capacity?: PositionCapacity;
}
